feat: integrate rss3 features

UncleBill opened the pull request:
successor of #4921

closes #4921

septs commented on packages/mask/src/plugins/Profile/SNSAdaptor/ConnectRSS3.tsx:
```diff
@@ -0,0 +1,136 @@
+                                <div>
+                                    <Typography className={classes.msg} variant="body2" gutterBottom>
+                                        By clicking the button above, you agree to the{' '}
+                                        <Link href="https://rss3.io/#privacy" target="_blank">
```
```suggestion
                                        <Link href="https://rss3.io#privacy" target="_blank">
```

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useFootprints.ts:
```diff
@@ -0,0 +1,23 @@
+    const [footprints, setFootprints] = useState<GeneralAssetWithTags[]>([])
+    const [isLoading, setIsLoading] = useState(true)
+
+    useEffect(() => {
```
Refactor to use [`useAsync`](https://github.com/streamich/react-use/blob/master/docs/useAsync.md).

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useDonations.ts:
```diff
@@ -0,0 +1,23 @@
+    const [donations, setDonations] = useState<GeneralAssetWithTags[]>([])
+    const [isLoading, setIsLoading] = useState(true)
+
+    useEffect(() => {
```
Refactor to use [`useAsync`](https://github.com/streamich/react-use/blob/master/docs/useAsync.md).

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/components/FootprintCard.tsx:
```diff
@@ -0,0 +1,65 @@
+}
+
+const formatDate = (ts: string): string => {
+    return new Date(Number.parseInt(ts, 16) * 1000).toLocaleDateString('en-US')
```
import fromUnixTime from 'date-fns/fromUnixTime'

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/common/rss3.ts:
```diff
@@ -0,0 +1,204 @@
+        composed: true,
+    })
+    document.dispatchEvent(evt)
+    console.log(event, detail)
```
```suggestion
```

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/common/address.ts:
```diff
@@ -0,0 +1,8 @@
+const formatter = (address: string | undefined): string => {
```
It duplicated `fromatEthereumAddress`.

UncleBill replied:
It's unused. Removed

wenluomask commented:
1. Since I just recovered my backup but didn't connect any wallet, it supposed to get an address from NFT avatar and display that address' NFTs. In this case, Web3 tab didn't display anything.(I tried to add an ens in bio and then delete later, Web3 can fetch avatar's address);
<img width="1038" alt="Screen Shot 2021-12-19 at 10 47 03 PM" src="https://user-images.githubusercontent.com/85149605/146679577-040e1443-8d6f-48ca-b942-c5059d3cf567.png">

2. Only ens can relate to RSS3 account. If I used address that registered in RSS3, it didn't recognize it is registered in RSS3. The address should be the relation of Mask, Twitter and RSS3, but not ens;

3. The current display of address should be the same format in three tabs;

4. "Contribution" cannot display fully, hopefully you can adjust the UI(Although the UI is as same as RSS3);
<img width="699" alt="Screen Shot 2021-12-19 at 11 26 35 PM" src="https://user-images.githubusercontent.com/85149605/146681041-c9dd8bd9-6d89-4983-9ac8-31891183d04b.png">

5. Attended should be capital;
<img width="730" alt="Screen Shot 2021-12-19 at 11 31 58 PM" src="https://user-images.githubusercontent.com/85149605/146681149-ae7e4460-e3b2-4641-8940-a19bb6d09928.png">

UncleBill commented:
> 1. Since I just recovered my backup but didn't connect any wallet, it supposed to get an address from NFT avatar and display that address' NFTs. In this case, Web3 tab didn't display anything.(I tried to add an ens in bio and then delete later, Web3 can fetch avatar's address);

> 
> <img alt="Screen Shot 2021-12-19 at 10 47 03 PM" width="1038" src="https://user-images.githubusercontent.com/85149605/146679577-040e1443-8d6f-48ca-b942-c5059d3cf567.png">
> 
> 2. Only ens can relate to RSS3 account. If I used address that registered in RSS3, it didn't recognize it is registered in RSS3. The address should be the relation of Mask, Twitter and RSS3, but not ens;

Now we can resolve use address from rss3 profile link

> 3. The current display of address should be the same format in three tabs;
Fixed
> 4. "Contribution" cannot display fully, hopefully you can adjust the UI(Although the UI is as same as RSS3);

> 
> <img alt="Screen Shot 2021-12-19 at 11 26 35 PM" width="699" src="https://user-images.githubusercontent.com/85149605/146681041-c9dd8bd9-6d89-4983-9ac8-31891183d04b.png">
> 

"Contrib" is not a typo, you can check out https://joshua.rss3.bio/gitcoins

> 5. Attended should be capital;
> 
> <img alt="Screen Shot 2021-12-19 at 11 31 58 PM" width="730" src="https://user-images.githubusercontent.com/85149605/146681149-ae7e4460-e3b2-4641-8940-a19bb6d09928.png">

Fixed

guanbinrui commented on packages/mask/src/plugins/Profile/services.ts:
```diff
@@ -0,0 +1,54 @@
+export function getFootprints(account: string): Promise<Response> {
+    const url = urlcat('https://hub.pass3.me/assets/list', {
+        personaID: account,
+        type: 'POAP',
```
An enum for this.

```ts
enum AssetType {
    NFT = 'NFT',
    GitcoinDonation = 'Gitcoin-Donation',
    POAP = 'POAP',
}
```

guanbinrui commented on packages/mask/src/plugins/Profile/apis/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './rss3'
```
Remove this file.

guanbinrui commented:
@UncleBill Please remove all of the unused files.

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/DonationsPage.tsx:
```diff
@@ -1,11 +1,74 @@
+                        rel="noopener noreferrer">
+                        <DonationCard
+                            imageUrl={donation.info.image_preview_url || RSS3_DEFAULT_IMAGE}
+                            name={donation.info.title || 'Inactive Project'}
```
i18n

guanbinrui commented on packages/mask/src/plugins/Profile/services.ts:
```diff
@@ -0,0 +1,52 @@
+        type: AssetType.GitcoinDonation,
+    })
+
+    return fetchJSON(url) as Promise<Response>
```
```suggestion
    return fetchJSON<Response>(url)
```

guanbinrui commented on packages/mask/src/plugins/Profile/services.ts:
```diff
@@ -0,0 +1,52 @@
+        type: AssetType.POAP,
+    })
+
+    return fetchJSON(url) as Promise<Response>
```
```suggestion
    return fetchJSON<Response>(url)
```

guanbinrui commented on packages/mask/src/plugins/Profile/services.ts:
```diff
@@ -0,0 +1,52 @@
+        return ''
+    }
+    const url = urlcat('https://rss3.domains/name/:id', { id: matched[1] })
+    const rsp = (await fetchJSON(url)) as NameInfo
```
```suggestion
    const rsp = (await fetchJSON<NameInfo>(url))
```

guanbinrui commented on packages/mask/src/plugins/Profile/services.ts:
```diff
@@ -0,0 +1,52 @@
+
+export async function getRss3Profile(address: string) {
+    const url = urlcat('https://hub.pass3.me/:address', { address })
+    const rsp = (await fetchJSON(url)) as RSS3Info
```
```suggestion
    const rsp = (await fetchJSON<RSS3Info>(url))
```

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useAddressByRss3ProfileLink.ts:
```diff
@@ -0,0 +1,24 @@
+import { useAsync } from 'react-use'
```
Rename this file by the name of the hook in it.

guanbinrui commented on packages/mask/src/plugins/Profile/SNSAdaptor/hooks/useAddressByRss3ProfileLink.ts:
```diff
@@ -0,0 +1,24 @@
+
+export function useAddressByRss3(profileUrl: string, nickname: string) {
+    const { value: address, loading } = useAsync(async () => {
+        let rss3id: string
+        let matched = nickname.match(RSS3_RNS_RE)
+        if (matched) {
+            rss3id = matched[1]
+        } else {
+            matched = profileUrl.match(RSS3_PROFILE_URL_RE)
+            rss3id = matched ? matched[1] : ''
+        }
```
```suggestion
        const matched = nickname.match(RSS3_RNS_RE) || profileUrl.match(RSS3_PROFILE_URL_RE)
        const id = matched ? matched[1] : ''
```

septs commented on packages/mask/src/plugins/Profile/constants.ts:
```diff
@@ -3,3 +3,7 @@ export const PLUGIN_NAME = 'Profile'
 
 export const RSS3_DEFAULT_IMAGE = 'https://rss3.mypinata.cloud/ipfs/QmVFq9qimnudPcs6QkQv8ZVEsvwD3aqETHWtS5yXgdbYY5'
+
+export const RSS3_PROFILE_URL_RE = /^https?:\/\/(\w+)\.rss3\.bio/
```
suggest, use named groups?

```suggestion
export const RSS3_PROFILE_URL_RE = /^https?:\/\/(?<name>\w+)\.rss3\.bio/
```

```
> pattern = /^https?:\/\/(?<name>\w+)\.rss3\.bio/
/^https?:\/\/(?<name>\w+)\.rss3\.bio/
> matched = pattern.exec('https://cz_binance.rss3.bio/')
[
  'https://cz_binance.rss3.bio',
  'cz_binance',
  index: 0,
  input: 'https://cz_binance.rss3.bio/',
  groups: [Object: null prototype] { name: 'cz_binance' }
]
> matched.groups.name
'cz_binance'
```

UncleBill replied:
iOS doesn't support regexp group yet.

septs replied:
> iOS doesn't support regexp group yet.

Safari 15.2 supported

Jack-Works replied:
![image](https://user-images.githubusercontent.com/5390719/146718892-7b41332f-c937-4f9f-9ec8-e3e735e16ff0.png)

All green according to https://kangax.github.io/compat-table/es2016plus/

septs commented on packages/mask/src/plugins/Profile/constants.ts:
```diff
@@ -1,3 +1,9 @@
+
+export const RSS3_PROFILE_URL_RE = /^https?:\/\/(\w+)\.rss3\.bio/
+
+export const RSS3_RNS_RE = /^(\w+)\.rss3$/
```
```suggestion
export const RSS3_RNS_RE = /^(?<name>\w+)\.rss3$/
```

UncleBill replied:
ditto

septs commented on packages/mask/background/services/helper/fetch.ts:
```diff
@@ -3,7 +3,8 @@ export async function fetch(url: string) {
 }
 
-export async function fetchJSON(url: string): Promise<unknown> {
+export async function fetchJSON<T extends unknown>(url: string): Promise<T> {
```
plz sync develop branch

wenluomask commented:
https://mask.atlassian.net/jira/software/c/projects/MASK/boards/20?modal=detail&selectedIssue=MASK-314

septs commented on packages/mask/src/plugins/Profile/SNSAdaptor/components/AddressViewer.tsx:
```diff
@@ -0,0 +1,90 @@
+        <div style={{ textAlign: 'left' }}>
+            <Typography variant="body2">{t('plugin_profile_binding_rules_title')}</Typography>
+            <ul className={classes.tipList}>
+                {rulesTipMap.map((item, index) => {
+                    return <li key={index}>{item}</li>
+                })}
```
```suggestion
                {rulesTipMap.map((item, index) => <li key={index}>{item}</li>)}
```

septs commented on packages/mask/src/plugins/Wallet/services/domain.ts:
```diff
@@ -1,22 +1,141 @@
+
+function getRSS3Id(nickname: string, profileURL: string) {
+    const matched = nickname.match(RSS3_RNS_RE) || profileURL.match(RSS3_PROFILE_URL_RE)
+    return matched ? matched.groups?.name ?? '' : ''
```
```suggestion
    return matched?.groups?.name ?? ''
```

septs commented on packages/mask/src/plugins/Wallet/services/domain.ts:
```diff
@@ -1,22 +1,141 @@
+function getEthereumName(twitterId: string, nickname: string, bio: string) {
+    const [matched] = nickname.match(ENS_RE) ?? bio.match(ENS_RE) ?? []
+    if (matched) return matched
+    return twitterId && !twitterId.endsWith('.eth') ? `${twitterId}.eth` : ''
```
```suggestion
    return !twitterId?.endsWith('.eth') ? `${twitterId}.eth` : ''
```

guanbinrui replied:
I cannot accept this suggestion.

septs commented on packages/web3-shared/evm/types/index.ts:
```diff
@@ -609,15 +609,19 @@ export interface Transaction {
-    ENS = 'ENS',
-    UNS = 'UNS',
-    DNS = 'DNS',
+    ADDRESS = 1,
```
@guanbinrui this modify is safe?

guanbinrui replied:
It's safe.

wenluomask commented:
6. I put the RSS3 profile link to bio, it showed failed to find any address;
<img width="730" alt="Screen Shot 2021-12-21 at 1 57 11 PM" src="https://user-images.githubusercontent.com/85149605/146880491-17d1976b-7ffb-4e9d-8c4b-81824450ef3c.png">

7. I had NFT avatar, but didn't show my address in Web3;
<img width="680" alt="Screen Shot 2021-12-21 at 2 13 50 PM" src="https://user-images.githubusercontent.com/85149605/146880778-4ee9d335-0aad-49b2-bfd6-4ae05a310fc6.png">

MingMing had NFT avatar and address in bio, but didn't show any address either;
<img width="699" alt="Screen Shot 2021-12-21 at 2 14 25 PM" src="https://user-images.githubusercontent.com/85149605/146880957-e7a34ebc-83a3-4d7b-96e0-bf8b400be83f.png">

septs commented on packages/mask/src/extension/options-page/DashboardComponents/CollectibleList/index.tsx:
```diff
@@ -232,9 +233,16 @@ export function CollectibleListAddress(props: CollectibleListAddressProps) {
+export function CollectionList({
+    address,
+    selectedNetwork,
+}: {
+    address: string
+    selectedNetwork: Web3Plugin.NetworkDescriptor | null
+}) {
```
plz extract this type declare as named interface definition

wenluomask commented:
Checked.
